From 5d3a43a42683c49bb7bf808c44f19981541d76b1 Mon Sep 17 00:00:00 2001
From: Ahmet Oeztuerk <Ahmet.Oeztuerk@consol.de>
Date: Wed, 25 Feb 2026 14:00:10 +0100
Subject: [PATCH] use passed timeout argument instead of global timeouts for
 host/check

handle_host_check and handle_svc_check get a data pointer of type
nebstruct_host/service_check_data.

this struct has another field called 'object_ptr' , that points to the raw
host/service check struct

in a planned naemon feature, these host/service check structs will get their own
'check_timeout' field. this field is initialized to the default global value for
the service/host check timeouts, purely found within the naemon source
code.users can override this default global value by specifying another
check_timeout in the config. when a neb request is fired, it now uses this
check_timeout set in the host/service timeout

additionally the check_result struct now has a 'timeout' field, this is due to
timeouts being customizable. it makes sense to add them to a check_result

first change is for mod_gearman to stop writing timeout in the gearman data
wrongly. instead of looking into nebstruct_host/service_data 'timeout' field it
was using global host/service timeouts from the naemon source code, which it is
including. after this data is written and sent out to a gearman queue, a worker
parses the lines for available fields, including 'timeout'

second change is for setting the check_result->timeout to 0 on orphaned checks,
just making sure that its zeroed since it might have a random value.

Dont merge it before changes in naemon.
---
 neb_module_naemon/mod_gearman.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/neb_module_naemon/mod_gearman.c b/neb_module_naemon/mod_gearman.c
index da9f713..afd85f0 100644
--- a/neb_module_naemon/mod_gearman.c
+++ b/neb_module_naemon/mod_gearman.c
@@ -41,8 +41,6 @@ extern unsigned long  event_broker_options;
 #define my_free nm_free
 extern int            currently_running_host_checks;
 extern int            currently_running_service_checks;
-extern int            service_check_timeout;
-extern int            host_check_timeout;
 extern int            process_performance_data;
 extern int            log_notifications;
 extern int            log_global_notifications;
@@ -972,7 +970,7 @@ static int handle_host_check( int event_type, void *data ) {
               target_queue,
               hst->name,
               timeval2double(&core_time) - hostdata->latency, // can only assume planned start date since next_check already advanced to next check and last_check still points to previous check
-              host_check_timeout,
+              hostdata->timeout,
               processed_command
             );
 
@@ -1020,6 +1018,7 @@ static int handle_host_check( int event_type, void *data ) {
         chk_result->check_type          = HOST_CHECK_ACTIVE;
         chk_result->start_time.tv_sec   = (unsigned long)time(NULL);
         chk_result->finish_time.tv_sec  = (unsigned long)time(NULL);
+        chk_result->timeout             = 0;
         chk_result->latency             = 0;
         mod_gm_add_result_to_list( chk_result );
         chk_result = NULL;
@@ -1123,7 +1122,7 @@ static int handle_svc_check( int event_type, void *data ) {
               svcdata->host_name,
               svcdata->service_description,
               timeval2double(&core_time) - svcdata->latency, // can only assume planned start date since next_check already advanced to next check and last_check still points to previous check
-              service_check_timeout,
+              svcdata->timeout,
               processed_command
             );
 
@@ -1178,6 +1177,7 @@ static int handle_svc_check( int event_type, void *data ) {
         chk_result->start_time.tv_sec   = (unsigned long)time(NULL);
         chk_result->finish_time.tv_sec  = (unsigned long)time(NULL);
         chk_result->latency             = 0;
+        chk_result->timeout             = 0;
         mod_gm_add_result_to_list( chk_result );
         chk_result = NULL;
     }
